src/Database/LSMTree/Internal/Lookup.hs

```diff
@@ -1,32 +1,165 @@
+      -- The result vector is not guaranteed to be full, so we slice it. This
+      -- means it might retain memory that is not used by the vector.
+      pure $ VUM.slice 0 resix res'
+  where
+    !rsN = V.length rs
+    !ksN = V.length ks
+
+    go :: forall s.
+         VUM.MVector s (RunIx, KeyIx)
+      -> PrimVar s ResIx
+      -> ST s (VUM.MVector s (RunIx, KeyIx))
+    go res0 resixVar = loop1 res0 0
+      where
+        -- Loop over all run indexes
+        loop1 ::
+             VUM.MVector s (RunIx, KeyIx)
+          -> RunIx
+          -> ST s (VUM.MVector s (RunIx, KeyIx))
+        loop1 !res1 !rix
+          | rix == rsN = pure res1
+          | otherwise
+          = do
+              res1' <- loop2 res1 0 (rlvBloom $ rs `V.unsafeIndex` rix)
+              loop1 res1' (rix+1)
+          where
+            -- Loop over all key indexes
+            loop2 ::
+                 VUM.MVector s (RunIx, KeyIx)
+              -> KeyIx
+              -> Bloom SerialisedKey
+              -> ST s (VUM.MVector s (RunIx, KeyIx))
+            loop2 !res2 !kix b
+              | kix == ksN = pure res2
+              | let !k = ks `V.unsafeIndex` kix
+              , Bloom.elem k b = do
+                  resix <- readPrimVar resixVar
+                  -- Grows the vector if we've reached the end. TODO: tune how
+                  -- much much we grow the vector each time based on the
+                  -- expected true- and false-positives.
+                  res2' <- if resix == VUM.length res2
+                            then VUM.unsafeGrow res2 ksN
+                            else pure res2
+                  VUM.unsafeWrite res2 resix (rix, kix)
+                  writePrimVar resixVar $! (resix +1)
+                  loop2 res2' (kix+1) b
+              | otherwise = loop2 res2 (kix+1) b
+
+-- | Perform a batch of fence pointer index searches, and create an 'IOOp' for
+-- each search result. The resulting vector has the same length as the
+-- @VU.Vector (RunIx, KeyIx)@ argument, because index searching always returns a
+-- positive search result.
+indexSearches ::
+     forall m h. PrimMonad m
+  => V.Vector (RunLookupView (Handle h))
+  -> V.Vector SerialisedKey
+  -> VU.Vector (RunIx, KeyIx) -- ^ Result of 'bloomQueries'
+  -> m (V.Vector (IOOp m h))
+indexSearches rs ks rkixs = do
+    -- The result vector has exactly the same length as @rkixs@.
+    res <- VM.unsafeNew n
+    go res
+    V.unsafeFreeze res
+  where
+    !n = VU.length rkixs
+
+    go :: VM.MVector (PrimState m) (IOOp m h) -> m ()
+    go res = loop 0
+      where
+        -- Loop over all indexes in @rkixs@
+        loop :: Int -> m ()
+        loop !i
+          | i == n = pure ()
+          | otherwise = do
+              let (!rix, !kix) = rkixs `VU.unsafeIndex` i
+                  !r     = rs `V.unsafeIndex` rix
+                  !k     = ks `V.unsafeIndex` kix
+                  !pspan = Index.toPageSpan $ Index.search k (rlvIndex r)
```

Comment:
We might want to have a specialised version of `Index.search` that returns the combo of the start page and number of pages (or first and last page) as a single-constructor result, so that we don't have to go via an intermediate (allocated) 2-constructor type.

Or maybe that could be the primary api for lookup since it covers both single and multi-page results.

Reply:
`PageSpan` now replaces `SearchResult`!